### ISSUE TITLE
Upload Directory

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#5117]
If we manage to get [Nanny Plugins](https://github.com/dask/distributed/issues/5116) in we might then consider a nanny plugin that uploads a directory to the worker and then optionally restarts that worker if necessary (because maybe we're shipping up a bundle of software).

When experimenting with this I found [this stack overflow post](https://stackoverflow.com/questions/1855095/how-to-create-a-zip-archive-of-a-directory) valuable.  It contains the following code that uses the `zip` module.

```python
import os
import zipfile
    
def zipdir(path, ziph):
    # ziph is zipfile handle
    for root, dirs, files in os.walk(path):
        for file in files:
            ziph.write(os.path.join(root, file), 
                       os.path.relpath(os.path.join(root, file), 
                                       os.path.join(path, '..')))
      
zipf = zipfile.ZipFile('Python.zip', 'w', zipfile.ZIP_DEFLATED)
zipdir('tmp/', zipf)
zipf.close()
```

## Special case of software

I personally would find this valuable when it comes to uploading a directory of local software.  In this situation there are a few considerations:

1.  We're going to want to restart the worker if we're uploading this on to running workers
2. It is probably also wise to exclude paths tha contain `.git/` in them.  Doing this I think that it should be easy to upload full packages of in-development software, which is a common pain point among folks trying out their own code.
3. We will want to append the path of the directory to the python path of the worker before we run any code.  In particular I wanted to do this with `dask` and so I wouldn't even be able to `import distributed`.  Maybe it's best to handle this with the PYTHONPATH environment variable

**Comment by mrocklin** [comment]
Here is a test that might make sense?

```python
from distributed.utils_test import gen_cluster
from dask.distributed import Nanny

@gen_cluster(client=True, Worker=Nanny)
async def test_upload_directory(c, s, a, b, tmp_path):
    with open(tmp_path / "foo.py", 'w') as f:
        f.write("x = 123")
    with open(tmp_path / "bar.py", 'w') as f:
        f.write("from foo import x")

    await c.upload_directory(tmp_path, restart=True)

    def f():
        import bar
        return bar.x

    results = await c.run(f)
    assert f[a.address] == 123
    assert f[b.address] == 123
```

**Comment by mrocklin** [comment]
I've implemented this in #5118

**Comment by jrbourbeau** [comment]
Closed via https://github.com/dask/distributed/pull/5118
